targets/f7/furi_hal/furi_hal_serial.c

```diff
@@ -0,0 +1,797 @@
+#include <furi_hal_serial.h>
+#include "furi_hal_serial_types_i.h"
+
+#include <stdbool.h>
+#include <stm32wbxx_ll_lpuart.h>
+#include <stm32wbxx_ll_usart.h>
+#include <stm32wbxx_ll_rcc.h>
+#include <stm32wbxx_ll_dma.h>
+#include <furi_hal_resources.h>
+#include <furi_hal_interrupt.h>
+#include <furi_hal_bus.h>
+
+#include <furi.h>
+
+#define FURI_HAL_USART_OVERSAMPLING LL_USART_OVERSAMPLING_16
+
+typedef struct {
+    uint8_t* buffer_rx_ptr;
+    size_t buffer_rx_index_write;
+    size_t buffer_rx_index_read;
+    bool enabled;
+    FuriHalSerialHandle* handle;
+    FuriHalSerialAsyncRxCallback rx_byte_callback;
+    FuriHalSerialDmaRxCallback rx_dma_callback;
+    void* context;
+} FuriHalSerial;
+
+static FuriHalSerial furi_hal_serial[FuriHalSerialIdMax] = {0};
+
+static size_t furi_hal_serial_dma_bytes_available(FuriHalSerialId ch);
+
+static void furi_hal_serial_async_rx_configure(
+    FuriHalSerialHandle* handle,
+    FuriHalSerialAsyncRxCallback callback,
+    void* context);
+
+static void furi_hal_serial_usart_irq_callback() {
+    FuriHalSerialRxEvent event = 0;
+    //if no errors
+    if(USART1->ISR & USART_ISR_RXNE_RXFNE) {
+        event |= FuriHalSerialRxEventData;
+    }
+    if(USART1->ISR & USART_ISR_IDLE) {
+        USART1->ICR = USART_ICR_IDLECF;
+        event |= FuriHalSerialRxEventIdle;
+    }
+    //if errors
+    if(USART1->ISR & USART_ISR_ORE) {
+        USART1->ICR = USART_ICR_ORECF;
+        event |= FuriHalSerialRxEventOverrunError;
+    }
+    if(USART1->ISR & USART_ISR_NE) {
+        USART1->ICR = USART_ICR_NECF;
+        event |= FuriHalSerialRxEventNoiseError;
+    }
+    if(USART1->ISR & USART_ISR_FE) {
+        USART1->ICR = USART_ICR_FECF;
+        event |= FuriHalSerialRxEventFrameError;
+    }
+    if(USART1->ISR & USART_ISR_PE) {
+        USART1->ICR = USART_ICR_PECF;
+        event |= FuriHalSerialRxEventFrameError;
+    }
+
+    if(furi_hal_serial[FuriHalSerialIdUsart].buffer_rx_ptr == NULL) {
```

Comment:
minor speedup: 
`const ...* usart_serial = &furi_hal_serial[FuriHalSerialIdUsart];`
then
`usart_serial->buffer_rx_ptr` etc etc

reason: does not index the array on each access

Reply:
will be refactored